### PR TITLE
Added SPSS weighting to the CHAID Algorithm

### PR DESCRIPTION
**Pull request by Rambatino**
Estimating weighted counts using this formula from SPSS Tree Algorithms:

image

### CODE REVIEW

File: CHAID/CHAID.py [hunk 1]
```diff
@@ -366,37 +369,63 @@ def node(self, rows, ind, dep, depth=0, parent=None, parent_decisions=None):
...
                     keys = set(col1_freq.keys()).union(col2_freq.keys())
 
-                    cr_table = [
+                    n_ij = np.array([
                         [col1_freq.get(k, 0) for k in keys],
                         [col2_freq.get(k, 0) for k in keys]
-                    ]
+                    ])
+
+                    if wt is not None:
```
Comment (xulaus) [inline CHAID/CHAID.py]: `is not None` not needed

Comment (Rambatino) [inline CHAID/CHAID.py]:
``` python 
>                   if wt:
E                   ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

File: CHAID/CHAID.py [hunk 2]
```diff
@@ -366,37 +369,63 @@ def node(self, rows, ind, dep, depth=0, parent=None, parent_decisions=None):
...
+            wt_freq = {}
             for col in unique:
                 counts = np.unique(dep.arr[index.arr == col], return_counts=True)
-                frequencies[col] = cl.defaultdict(int)
-                frequencies[col].update(np.transpose(counts))
+                freq[col] = cl.defaultdict(int)
+                wt_freq[col] = cl.defaultdict(int)
+                freq[col].update(np.transpose(counts))
+                if wt is not None:
+                    for dep_v in set(dep.arr):
```
Comment (Rambatino) [inline CHAID/CHAID.py]: I know. Will look into optimising it now

File: CHAID/CHAID.py [hunk 3]
```diff
@@ -366,37 +369,63 @@ def node(self, rows, ind, dep, depth=0, parent=None, parent_decisions=None):
...
+
+                        alpha, beta, eps = (1, 1, 1)
+                        while eps > 0.000001:
+                            alpha = alpha * np.vstack(n_ij.sum(axis=1) / m_ij.sum(axis=1))
+                            beta = n_ij.sum(axis=0) / (alpha * w_ij).sum(axis=0)
+                            eps = np.max(np.absolute(w_ij * alpha * beta - m_ij))
+                            m_ij = w_ij * alpha * beta
+                        n_ij = m_ij
+
+                    exp = (np.vstack(n_ij.sum(axis=1)) * n_ij.sum(axis=0)) / n_ij.sum().astype(float)
```
Comment (Rambatino) [inline CHAID/CHAID.py]: I don't think we need vstack here on the assumption that the ndarray is always square. Is it safe to make that assumption?

Comment (Rambatino) [inline CHAID/CHAID.py]: Nope, if you have 3 types of dependent variable categories, you'll get non square ndarrays

**Comment by xulaus** [comment]
Sorry about the weird not line note that was ment to be a line note. Reviewing on a phone is hard :(.

File: CHAID/CHAID.py (hunk 1 above)
Comment (xulaus) [inline CHAID/CHAID.py]: My bad

File: CHAID/CHAID.py [hunk 4]
```diff
@@ -337,7 +338,8 @@ def node(self, rows, ind, dep, depth=0, parent=None, parent_decisions=None):
             self.node_count += 1
             return self.tree_store
 
-        split = self.generate_best_split(ind, dep)
+        wt = (self.weights[rows] if self.weights else None)
```
Comment (xulaus) [inline CHAID/CHAID.py]: This is the line that was better as a if statement with the not none check.

File: CHAID/CHAID.py [hunk 5]
```diff
@@ -366,46 +377,85 @@ def node(self, rows, ind, dep, depth=0, parent=None, parent_decisions=None):
...
+
+                    dof = (n_ij.shape[0] - 1) * (n_ij.shape[1] - 1)
+                    ret = st_chi(n_ij, f_exp=m_ij, ddof=n_ij.size - 1 - dof, axis=None)
+
+                    sub_data[j] = (comb, ret[1], ret[0])
+
+                choice, highest_p_join, chi_join = max(sub_data, key=lambda x: (x[1], x[2]))
+
+                if highest_p_join < self.alpha_merge:
+                    if len(freq.keys()) == 2 or len(wt_freq.keys()) == 2: # have actual p-value
```
Comment (xulaus) [inline CHAID/CHAID.py]: I'm not sure this branch really adds much in the way of speed, and it makes the code harder to follow. I'd remove it.

File: CHAID/CHAID.py [hunk 6]
```diff
@@ -431,14 +481,31 @@ def generate_best_split(self, ind, dep):
...
                 unique.remove(choice[1])
 
-                for val, count in frequencies[choice[1]].items():
-                    frequencies[choice[0]][val] += count
-                del frequencies[choice[1]]
+                for val, count in freq[choice[1]].items():
+                    freq[choice[0]][val] += count
+                del freq[choice[1]]
+
+                if wt is not None:
```
Comment (xulaus) [inline CHAID/CHAID.py]: There is a lot of testing against weights existing. Can we just set all the weights to 1 in the unweighted case? Or is that too slow?

Comment (Rambatino) [inline CHAID/CHAID.py]: HA! Making everyone the same weight isn't the same as not using weights. Retarded, I know.

**Comment by Rambatino** [comment]
@xulaus can you unblock the merge. Have removed the if clause (although this could add quite a few hundred thousand calculations per tree generation (although it isn't a bottleneck)) and have added the spec

File: CHAID/CHAID.py (hunk 6 above)
Comment (samwood1) [inline CHAID/CHAID.py]: Could you check this assumption with Markus? It seems odd.

**Comment by Rambatino** [comment]
Sorry @xulaus we can talk when you're back but I'm merging this.

**Pull request merged**